# Pandas Basics — Part 3

In this lesson, we're going to introduce more fundamentals of [Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/overview.html), a powerful Python library for working with tabular data like CSV files.

We will review skills learned from the last two lessons and introduce how to:

- Check for duplicate data
- Clean and transform data
- Manipulate string data
- Apply functions
- Reset index
- Bonus! Create an interactive data viz

___

## 1. Dataset - *The Pudding*'s Film Dialogue Data

<blockquote class="epigraph" style=" padding: 10px">

Lately, Hollywood has been taking so much shit for rampant sexism and racism. The prevailing theme: white men dominate movie roles.

But it’s all rhetoric and no data, which gets us nowhere in terms of having an informed discussion. How many movies are actually about men? What changes by genre, era, or box-office revenue? What circumstances generate more diversity?

-Hannah Anderson and Matt Daniels, ["Film Dialogue from 2,000 screenplays, Broken Down by Gender and Age"](https://pudding.cool/2017/03/film-dialogue/)

</blockquote>


The dataset that we're working with in this lesson is taken from Hannah Andersen and Matt Daniels's *Pudding* essay, ["Film Dialogue from 2,000 screenplays, Broken Down by Gender and Age"](https://pudding.cool/2017/03/film-dialogue/). The dataset provides information about 2,000 films from 1925 to 2015, including characters’ names, genders, ages, how many words each character spoke in each film, the release year of each film, and how much money the film grossed. They included character gender information because they wanted to contribute data to a broader conversation about how "white men dominate movie roles."

Yet transforming complex social constructs like gender into quantifiable data is tricky and historically fraught. They claim, in fact, that one of the [most frequently asked questions](https://medium.com/@matthew_daniels/faq-for-the-film-dialogue-by-gender-project-40078209f751) about the piece is about gender: “Wait, but let’s talk about gender. How do you know the monster in Monsters Inc. is a boy!" The short answer is that they don't. To determine character gender, they used actors' IMDB information, which they acknowledge is an imperfect approach: "Sometimes, women voice male characters. Bart Simpson, for example, is voiced by a woman. We’re aware that this means some of the data is wrong, AND we’re still fine with the methodology and approach."

As we work with this data, we want to be critical and cognizant of this approach to gender. How does such a binary understanding of gender, gleaned from the IMDB pages of actors, influence our later results and conclusions? What do we gain by using such an approach, and what do we lose? How else might we have encoded or determined gender for the same data? 

___

## 2. Setup Our Project

### 2.1 Import Pandas

To use the Pandas library, we first need to `import` it.

In [1]:
import pandas as pd

The above `import` statement not only imports the Pandas library but also gives it an alias or nickname — `pd`. This alias will save us from having to type out the entire words `pandas` each time we need to use it. Many Python libraries have commonly used aliases like `pd`.

### 2.2 Set Display Settings

By default, Pandas will display 60 rows and 20 columns. I often change [Pandas' default display settings](https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html) to show more rows or columns.

In [2]:
pd.options.display.max_rows = 100

### 2.3 Read in CSV File

To read in a CSV file, we will use the function `pd.read_csv()` and insert the name of our desired file path. 

In [3]:
film_df = pd.read_csv('../data/Pudding/Pudding-Film-Dialogue-Salty.csv', delimiter=",", encoding='utf-8')

When reading in the CSV file, we also specified the `encoding` and `delimiter`. The `delimiter` parameter specifies the character that separates or "delimits" the columns in our dataset. For CSV files, the delimiter will most often be a comma. (CSV is short for *Comma Separated Values*.) Sometimes, however, the delimiter of a CSV file might be a tab (`\t`) or, more rarely, another character.

### 2.4 Display Data

We can display a DataFrame in a Jupyter notebook simply by running a cell with the variable name of the DataFrame.

<div class="admonition pandasreview" name="html-admonition" style="background: black; color: white; padding: 10px">
<p class="title">Pandas Review</p>
<code>NaN</code> is the Pandas value for any missing data. See <a href="https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html?highlight=nan/">"Working with missing data"</a> for more information.
        
</div>

In [4]:
film_df

script_id imdb_character_name  words gender   age    imdb_id  \
0            280               betty    311      f  35.0  tt0112579   
1            280     carolyn johnson    873      f   NaN  tt0112579   
2            280             eleanor    138      f   NaN  tt0112579   
3            280     francesca johns   2251      f  46.0  tt0112579   
4            280               madge    190      f  46.0  tt0112579   
...          ...                 ...    ...    ...   ...        ...   
23047       9254             maurice   1107      m  71.0  tt0101414   
23048       9254     monsieur d'arqu    114      m  58.0  tt0101414   
23049       9254          mrs. potts    564      f  66.0  tt0101414   
23050       9254            wardrobe    121      f  54.0  tt0101414   
23051       9254          mrs. potts    564      f  66.0  tt0101414   

                               title  year  gross  proportion_of_dialogue  
0      The Bridges of Madison County  1995  142.0                0.048639  
1      The Bridges of Madison County  1995  142.0                0.136534  
2      The Bridges of Madison County  1995  142.0                0.021583  
3      The Bridges of Madison County  1995  142.0                0.352049  
4      The Bridges of Madison County  1995  142.0                0.029715  
...                              ...   ...    ...                     ...  
23047           Beauty and the Beast  1991  452.0                0.108967  
23048           Beauty and the Beast  1991  452.0                0.011222  
23049           Beauty and the Beast  1991  452.0                0.055517  
23050           Beauty and the Beast  1991  452.0                0.011911  
23051           Beauty and the Beast  1991  452.0                0.055517  

[23052 rows x 10 columns]

There are a few important things to note about the DataFrame displayed here:

* Index
    * The bolded ascending numbers in the very left-hand column of the DataFrame is called the Pandas *Index*. You can select rows based on the Index.
    * By default, the Index is a sequence of numbers starting with zero. However, you can change the Index to something else, such as one of the columns in your dataset.

* Truncation
    * The DataFrame is truncated, signaled by the ellipses in the middle `...` of every column.
    * The DataFrame is truncated because we set our default display settings to 100 rows. Anything more than 100 rows will be truncated. To display all the rows, we would need to alter Pandas' default display settings yet again.

* Rows x Columns
    * Pandas reports how many rows and columns are in this dataset at the bottom of the output (23,052 x 10 columns).

### 2.5 Display First *n* Rows

To look at the first *n* rows in a DataFrame, we can use a method called `.head()`.

In [21]:
film_df.head(10)

script_id imdb_character_name  words gender   age    imdb_id  \
0        280               betty    311      f  35.0  tt0112579   
1        280     carolyn johnson    873      f   NaN  tt0112579   
2        280             eleanor    138      f   NaN  tt0112579   
3        280     francesca johns   2251      f  46.0  tt0112579   
4        280               madge    190      f  46.0  tt0112579   
5        280     michael johnson    723      m  38.0  tt0112579   
6        280      robert kincaid   1908      m  65.0  tt0112579   
7        623        bobby korfin    328      m   NaN  tt0179626   
8        623     daphne handlova    409      f  28.0  tt0179626   
9        623     deputy chief fi    347      m   NaN  tt0179626   

                           title  year  gross  proportion_of_dialogue  
0  The Bridges of Madison County  1995  142.0                0.048639  
1  The Bridges of Madison County  1995  142.0                0.136534  
2  The Bridges of Madison County  1995  142.0                0.021583  
3  The Bridges of Madison County  1995  142.0                0.352049  
4  The Bridges of Madison County  1995  142.0                0.029715  
5  The Bridges of Madison County  1995  142.0                0.113075  
6  The Bridges of Madison County  1995  142.0                0.298405  
7                     15 Minutes  2001   37.0                0.036012  
8                     15 Minutes  2001   37.0                0.044906  
9                     15 Minutes  2001   37.0                0.038098

### 2.6 Display Random Sample

To look at a random sample of rows, we can use the `.sample()` method.

In [22]:
film_df.sample(10)

script_id imdb_character_name  words gender   age    imdb_id  \
3519        1483           auntie em    295      f  63.0  tt0032138   
5395        1991       holden mcneil   9462      m  25.0  tt0118842   
5238        1957     lightning mcque   1968      m  43.0  tt1216475   
2992        1350         tariq asani    267      m  31.0  tt0360009   
18281       6858                mara    568      f   NaN  tt0114852   
8333        2764            nurse #1    162      f  37.0  tt0264796   
13345       4394            jonathan   1168      m  47.0  tt1691920   
16889       6053              bomber    392      m   NaN  tt1532958   
15704       5297           mr. green    825      m  38.0  tt0088930   
75           642          kurt zagon    914      m  60.0  tt0094602   

                       title  year  gross  proportion_of_dialogue  
3519        The Wizard of Oz  1939  839.0                0.031426  
5395             Chasing Amy  1997   22.0                0.300973  
5238                  Cars 2  2011  209.0                0.155622  
2992                 Spartan  2004    6.0                0.043337  
18281  Village of the Damned  1995   18.0                0.099006  
8333         Life as a House  2001   23.0                0.007482  
13345        I Melt with You  2011    NaN                0.159041  
16889     Battle of the Year  2013    9.0                0.035617  
15704                   Clue  1985    7.0                0.087701  
75             Above the Law  1988   40.0                0.116522

### 2.7 Examine Data

**Shape**

To explicitly check for how many rows vs columns make up a dataset, we can use the `.shape` method.

In [23]:
film_df.shape

(23052, 10)

There are 23,052 rows and 10 columns.

### 2.8 Reminder: Data Types

Just like Python has different data types, Pandas has different data types, too. These data types are automatically assigned to columns when we read in a CSV file. We can check these Pandas data types with the [`.dtypes` method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html).



| **Pandas Data Type** |  **Explanation**                                                                                   |
|:-------------:|:---------------------------------------------------------------------------------------------------:|
| `object`         | string                                                                               |
| `float64`         | float                                               |
| `int64`       | integer                                                        |
| `datetime64`       |  date time              

In [24]:
film_df.dtypes

script_id                   int64
imdb_character_name        object
words                       int64
gender                     object
age                       float64
imdb_id                    object
title                      object
year                        int64
gross                     float64
proportion_of_dialogue    float64
dtype: object

It's important to always check the data types in your DataFrame. For example, sometimes numeric values will accidentally be interpreted as a string object. To perform calculations on this data, you would need to first convert that column from a string to an integer.

### 2.9 Columns

We can also check the column names of the DataFrame with [`.columns`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.columns.html)

In [25]:
film_df.columns

Index(['script_id', 'imdb_character_name', 'words', 'gender', 'age', 'imdb_id',
       'title', 'year', 'gross', 'proportion_of_dialogue'],
      dtype='object')

## 3. Summary Statistics &amp; Data Processing

Summary stats can help us spot initial patterns or even issues in the data.

In [26]:
film_df.describe(include='all')

script_id imdb_character_name         words gender           age  \
count   23052.000000               23050  23052.000000  23052  18266.000000   
unique           NaN               17613           NaN      3           NaN   
top              NaN              doctor           NaN      m           NaN   
freq             NaN                  36           NaN  16135           NaN   
mean     4195.580904                 NaN    907.870684    NaN     42.386839   
std      2473.518048                 NaN   1399.538583    NaN     59.714730   
min       280.000000                 NaN    101.000000    NaN      3.000000   
25%      2095.000000                 NaN    193.000000    NaN     30.000000   
50%      3694.000000                 NaN    396.000000    NaN     39.000000   
75%      6234.000000                 NaN    980.000000    NaN     50.000000   
max      9254.000000                 NaN  28102.000000    NaN   2013.000000   

          imdb_id      title          year         gross  \
count       23052      23052  23052.000000  19390.000000   
unique       2000       1994           NaN           NaN   
top     tt0116905  Lone Star           NaN           NaN   
freq           40         40           NaN           NaN   
mean          NaN        NaN   1998.133828    106.805570   
std           NaN        NaN     14.745342    145.992394   
min           NaN        NaN   1929.000000      0.000000   
25%           NaN        NaN   1992.000000     22.000000   
50%           NaN        NaN   2001.000000     56.000000   
75%           NaN        NaN   2009.000000    136.000000   
max           NaN        NaN   2015.000000   1798.000000   

        proportion_of_dialogue  
count             23052.000000  
unique                     NaN  
top                        NaN  
freq                       NaN  
mean                  0.086512  
std                   0.107740  
min                   0.001537  
25%                   0.019771  
50%                   0.042421  
75%                   0.104166  
max                   0.923422

**Read the data for any outliers, anomalies, or potential problems.**

Notice how the maximum value in the "age" column is 2013! That seems like an error.

In [27]:
film_df[film_df['age'] == 2013]

script_id imdb_character_name  words gender     age    imdb_id  \
11639       3737       lucas solomon    190      m  2013.0  tt0993846   

                         title  year  gross  proportion_of_dialogue  
11639  The Wolf of Wall Street  2013  125.0                0.011077

Let's drop this row from the dataset by using the `.drop()` method and the Index number of the row.

In [28]:
film_df = film_df.drop(11639)

Now if we look for it again, that row is gone.

In [29]:
film_df[film_df['age'] == 2013]

Empty DataFrame
Columns: [script_id, imdb_character_name, words, gender, age, imdb_id, title, year, gross, proportion_of_dialogue]
Index: []

### 3.1 Rename Columns

In [30]:
film_df = film_df.rename(columns={'imdb_character_name': 'character', 'year': 'release_year'})

In [31]:
film_df.head()

script_id        character  words gender   age    imdb_id  \
0        280            betty    311      f  35.0  tt0112579   
1        280  carolyn johnson    873      f   NaN  tt0112579   
2        280          eleanor    138      f   NaN  tt0112579   
3        280  francesca johns   2251      f  46.0  tt0112579   
4        280            madge    190      f  46.0  tt0112579   

                           title  release_year  gross  proportion_of_dialogue  
0  The Bridges of Madison County          1995  142.0                0.048639  
1  The Bridges of Madison County          1995  142.0                0.136534  
2  The Bridges of Madison County          1995  142.0                0.021583  
3  The Bridges of Madison County          1995  142.0                0.352049  
4  The Bridges of Madison County          1995  142.0                0.029715

### 3.2 Drop Columns

In [32]:
film_df = film_df.drop(columns='imdb_id')

### 3.3 Dealing with missing data .isna() / .notna()

Pandas has special ways of dealing with missing data. As you may have already noticed, blank rows in a CSV file show up as `NaN` in a Pandas DataFrame.

To filter and count the number of missing/not missing values in a dataset, we can use the special `.isna()` and `.notna()` methods on a DataFrame or Series object. 

The `.isna()` and `.notna()` methods return True/False pairs for each row, which we can use to filter the DataFrame for any rows that have information in a given column. 

In [33]:
film_df[film_df['character'].isna()]

script_id character  words gender  age           title  release_year  \
4656        1807       NaN    146      f  NaN         Beloved          1998   
19448       7445       NaN    520      f  NaN  House of Games          1987   

       gross  proportion_of_dialogue  
4656    42.0                0.005162  
19448    5.0                0.165184

This is important information for the sake of better understanding our dataset. But it's also important because `NaN` values are treated as *floats*, not *strings*. If we tried to manipulate this column as text data, we would get an error. For this reason, we're going to replace or "fill" these `NaN` values with the string "No Character Data" by using the `.fillna()` method.

In [34]:
film_df['character'] = film_df['character'].fillna('No Character Data')

In [35]:
film_df[film_df['character'].isna()]

Empty DataFrame
Columns: [script_id, character, words, gender, age, title, release_year, gross, proportion_of_dialogue]
Index: []

### 3.4 Check for Duplicates 

We can check for duplicate rows by using the `.duplicated()` method and setting the parameter `keep=False`, which will display all the duplicated values in the dataset — rather than just the first duplicated value `keep='first'` or the last duplicated value `keep='last'`.

In [36]:
film_df.duplicated(keep=False)

0        False
1        False
2        False
3        False
4        False
         ...  
23047    False
23048    False
23049     True
23050    False
23051     True
Length: 23051, dtype: bool

The output above is reporting whether each row in the dataset is a duplicate. We can use the `.duplicated()` method inside a filter to isolate only the rows in the dataframe that are exact duplicates.

In [37]:
film_df[film_df.duplicated(keep=False)]

script_id      character  words gender   age  \
21001       8099          c-3po    138      m  69.0   
21002       8099          c-3po    138      m  69.0   
21963       8560  kilgore trout   2673      m  63.0   
21964       8560  kilgore trout   2673      m  63.0   
22935       9222   groomsman #5    238      m   NaN   
22936       9222   groomsman #5    238      m   NaN   
23049       9254     mrs. potts    564      f  66.0   
23051       9254     mrs. potts    564      f  66.0   

                                            title  release_year  gross  \
21001  Star Wars: Episode VII - The Force Awakens          2015  927.0   
21002  Star Wars: Episode VII - The Force Awakens          2015  927.0   
21963                      Breakfast of Champions          1999    NaN   
21964                      Breakfast of Champions          1999    NaN   
22935                            Wedding Crashers          2005  283.0   
22936                            Wedding Crashers          2005  283.0   
23049                        Beauty and the Beast          1991  452.0   
23051                        Beauty and the Beast          1991  452.0   

       proportion_of_dialogue  
21001                0.020702  
21002                0.020702  
21963                0.197211  
21964                0.197211  
22935                0.012032  
22936                0.012032  
23049                0.055517  
23051                0.055517

We can drop duplicates from the DataFrame with the `.drop_duplicates()` method and choose to keep the first instance of the duplicate or the last instance.

In [38]:
film_df = film_df.drop_duplicates(keep='first')

Now if we check the data for duplicates again, they should be all gone.

In [39]:
film_df[film_df.duplicated(keep=False)]

Empty DataFrame
Columns: [script_id, character, words, gender, age, title, release_year, gross, proportion_of_dialogue]
Index: []

## 4. Pandas `.str` Methods

Remember all the special things that you can do with Python strings aka [string methods](https://melaniewalsh.github.io/Intro-Cultural-Analytics/Python/String-Methods.html)?

Pandas has special [Pandas string methods](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html#string-methods), too. Many of them are very similar to Python string methods, except they will transform every single string value in a column, and we have to add `.str` to the method chain.

| **Pandas String Method** | **Explanation**                                                                                   |
|:-------------:|:---------------------------------------------------------------------------------------------------:|
| df['column_name']`.str.lower()`         | makes the string in each row lowercase                                                                                |
| df['column_name']`.str.upper()`         | makes the string in each row uppercase                                                |
| df['column_name']`.str.title()`         | makes the string in each row titlecase                                                |
| df['column_name']`.str.replace('old string', 'new string')`      | replaces `old string` with `new string` for each row |
| df['column_name']`.str.contains('some string')`      | tests whether string in each row contains "some string" |
| df['column_name']`.str.split('delim')`          | returns a list of substrings separated by the given delimiter |
| df['column_name']`.str.join(list)`         | opposite of split(), joins the elements in the given list together using the string                                                                        |
                                                            

For example, to transform every character's name in the "character" column from lowercase to uppercase, we can use `.str.upper()` 

In [40]:
film_df['character'].str.upper()

0                  BETTY
1        CAROLYN JOHNSON
2                ELEANOR
3        FRANCESCA JOHNS
4                  MADGE
              ...       
23046            LUMIERE
23047            MAURICE
23048    MONSIEUR D'ARQU
23049         MRS. POTTS
23050           WARDROBE
Name: character, Length: 23047, dtype: object

To transform every character's name in the "character" column to lowercase, we can use `.str.lower()`

In [41]:
film_df['character'].str.lower()

0                  betty
1        carolyn johnson
2                eleanor
3        francesca johns
4                  madge
              ...       
23046            lumiere
23047            maurice
23048    monsieur d'arqu
23049         mrs. potts
23050           wardrobe
Name: character, Length: 23047, dtype: object

If we want to replace the gender columns's single letter abbreviation for "male" / "female" (sex) with "man" / "woman" (gender identity), we could use the `.str.replace()` method. 

In [42]:
film_df['gender'] = film_df['gender'].str.replace('m', 'man')

In [43]:
film_df['gender'] = film_df['gender'].str.replace('f', 'woman')

In [44]:
film_df.sample(10)

script_id        character  words gender   age  \
18526       7014            julie    256  woman  22.0   
10160       3289  charlie mccorry    584    man  40.0   
15578       5214              guy    163    man   NaN   
649          793            ellis    223    man   NaN   
892          851         peterson    207    man   NaN   
1190         915     clear rivers    472  woman  24.0   
3032        1360          lama su    350    man  48.0   
10153       3288         somerset   8838    man  58.0   
10421       3352       anna crowe    916  woman  31.0   
7132        2442             noah    160    man  32.0   

                                              title  release_year  gross  \
18526                                     Showgirls          1995   40.0   
10160                                 The Searchers          1956    NaN   
15578                            The Wedding Singer          1998  148.0   
649                                        Cellular          2004   44.0   
892                                       Deception          2008    5.0   
1190                              Final Destination          2000   86.0   
3032   Star Wars: Episode II - Attack of the Clones          2002  465.0   
10153                                         Se7en          1995  200.0   
10421                               The Sixth Sense          1999  503.0   
7132                            The Grapes of Wrath          1940    NaN   

       proportion_of_dialogue  
18526                0.023103  
10160                0.036878  
15578                0.011877  
649                  0.023130  
892                  0.024247  
1190                 0.163946  
3032                 0.046574  
10153                0.428281  
10421                0.080549  
7132                 0.006841

We can use the `.str.contains()` to search for particular words or phrases in a column, such as "Star Wars."

In [ ]:
film_df[film_df['title'].str.contains('Star Wars')].head()

Let's try "Mean Girls"

In [ ]:
film_df[film_df['title'].str.contains('Mean Girls')]

## 5. Applying Functions

With the `.apply()` method, we can run a function on every single row in a Pandas column or dataframe.

In [47]:
def make_text_title_case(text):
    title_case_text = text.title()
    return title_case_text

In [48]:
make_text_title_case("betty")

'Betty'

In [49]:
film_df['character'].apply(make_text_title_case)

0                  Betty
1        Carolyn Johnson
2                Eleanor
3        Francesca Johns
4                  Madge
              ...       
23046            Lumiere
23047            Maurice
23048    Monsieur D'Arqu
23049         Mrs. Potts
23050           Wardrobe
Name: character, Length: 23047, dtype: object

In [50]:
film_df['character'] = film_df['character'].apply(make_text_title_case)

In [51]:
film_df.sample(10)

script_id        character  words gender   age  \
21944       8559      Joey Pinero    288    man  40.0   
1698        1044   Anna Mcdoogles   1346  woman  37.0   
11675       3746  Randy 'The Ram'   5220    man  56.0   
801          821            Peter    804    man   NaN   
15636       5236  Michael Jackson    151    man   NaN   
1059         885      Lord Howard    154    man  65.0   
9011        2949  Cosmo Castorini   1196    man  67.0   
473          745  Deputy Da Ellis    462    man  37.0   
17745       6535      Rick Cowley    150    man  46.0   
2086        1140  Dr. Arthur Neum    267    man  50.0   

                           title  release_year  gross  proportion_of_dialogue  
21944                Primal Fear          1996  110.0                0.021510  
1698      The Invention of Lying          2009   21.0                0.160105  
11675               The Wrestler          2008   31.0                0.496009  
801                      Contact          1997  191.0                0.068449  
15636                    Willard          1971  101.0                0.016402  
1059   Elizabeth: The Golden Age          2007   20.0                0.021050  
9011                  Moonstruck          1987  179.0                0.063957  
473             The Black Dahlia          2006   29.0                0.040477  
17745              The Dark Half          1993   20.0                0.014219  
2086                    The Mask          1994  249.0                0.042995

## 6. Filter a DataFrame

We can filter the DataFrames for only characters who are men or women.

In [84]:
men_film_df = film_df[film_df['gender'] == 'man']

In [85]:
women_film_df = film_df[film_df['gender'] == 'woman']

## 7. Aggregate the data by grouping variables/columns with .groupby()

We can use the `.groupby()` function to group all the men characters in each film and sum up their total dialogue.

By adding a Python string slice, we can identify the top 20 films with the greatest proportion of men speaking.

```{margin} Line Breaks
If a line of code gets too long, you can create a line break with a backslash `\`
```

In [86]:
men_film_df.groupby('title')[['proportion_of_dialogue']]\
.sum().sort_values(by='proportion_of_dialogue', ascending=False)[:20]

proportion_of_dialogue
title                                                     
The Men Who Stare at Goats                             1.0
Kagemusha                                              1.0
The Wild Bunch                                         1.0
Killing Them Softly                                    1.0
Stalag 17                                              1.0
There Will Be Blood                                    1.0
Fury                                                   1.0
The Revenant                                           1.0
Saving Private Ryan                                    1.0
Crimson Tide                                           1.0
Die Hard: With a Vengeance                             1.0
Jeremiah Johnson                                       1.0
Conquest of the Planet of the Apes                     1.0
Duel                                                   1.0
The Hunt for Red October                               1.0
Platoon                                                1.0
Reservoir Dogs                                         1.0
The Battle of Algiers                                  1.0
Armored                                                1.0
Space Jam                                              1.0

We can use the `.groupby()` function to group all the women characters in each film and sum up their total dialogue.

By adding a Python string slice, we can identify the top 20 films with the greatest proportion of women speaking.

```{margin} Line Breaks
If a line of code gets too long, you can create a line break with a backslash `\`
```

In [87]:
women_film_df.groupby('title')[['proportion_of_dialogue']]\
.sum().sort_values(by='proportion_of_dialogue', ascending=False)[:20]

proportion_of_dialogue
title                                                 
The Descent                                   1.000000
Now and Then                                  1.000000
Precious                                      0.993541
Martyrs                                       0.965550
The Hand That Rocks the Cradle                0.933750
Agnes of God                                  0.922482
Heavenly Creatures                            0.919368
The Help                                      0.916947
3 Women                                       0.899530
The Watermelon Woman                          0.894676
Grandma                                       0.894137
The Craft                                     0.891866
Easy A                                        0.888131
The Others                                    0.879972
Black Christmas                               0.877467
Wadjda                                        0.870567
For Colored Girls                             0.870247
The Roommate                                  0.869879
Mrs. Winterbourne                             0.864730
Practical Magic                               0.856693

## 8. Reset Index

We can transform a Groupby object into a DataFrame with a regular Index by tacking on `.reset_index()`.

```{margin} Line Breaks
If a line of code gets too long, you can create a line break with a backslash `\`
```

In [88]:
women_film_df.groupby('title')[['proportion_of_dialogue']]\
.sum().sort_values(by='proportion_of_dialogue', ascending=False).reset_index()

title  proportion_of_dialogue
0                        The Descent                1.000000
1                       Now and Then                1.000000
2                           Precious                0.993541
3                            Martyrs                0.965550
4     The Hand That Rocks the Cradle                0.933750
...                              ...                     ...
1935                 The Last Castle                0.011139
1936               The Damned United                0.010909
1937                   Thirteen Days                0.010834
1938                  Men in Black 3                0.007812
1939               Full Metal Jacket                0.007170

[1940 rows x 2 columns]

## Bonus — Interactive Data Visualization

In [1]:
#Import necessary Bokeh modules
from bokeh.plotting import figure, show
from bokeh.models import ColumnDataSource, NumeralTickFormatter
from bokeh.io import output_notebook, show
from bokeh.palettes import RdBu
from bokeh.transform import linear_cmap, jitter

ModuleNotFoundError: No module named 'bokeh'

In [ ]:
#Set up Bokeh to work in Jupyter notebook
output_notebook()

Loading BokehJS ...

Here's an interactive data visualization of these Hollywood films by release year and percentage of women dialogue. This data viz was created with the Python library [Bokeh](https://docs.bokeh.org/en/latest/index.html), which we will discuss in some later lessons.

* Scroll to zoom in
* Hover to see more information about each point

To see the code that created this visualization, select "Click to show" below.

In [2]:
#Make groupby into a new DataFrame
dialogue_df = women_film_df.groupby(['title', 'release_year'])[['proportion_of_dialogue']].sum()\
.sort_values(by='proportion_of_dialogue', ascending=False).reset_index()

# Set up the source data that will suppply the x,y columns and the film title hover text
source = ColumnDataSource(dialogue_df)

# Set the hover tool tip to the film title, release year, and proportion of dialogue

TOOLTIPS = [("Title", "@title"),
            ("Year", "@release_year"),
           ("Women Dialogue", "@{proportion_of_dialogue}{%0.2f}")]

#Set up Bokeh plot with title, labels, 
bokeh_plot = figure(title="How Much Do Women Speak in Hollywood Films?", x_axis_label = 'Film Release Year',
                    y_axis_label = 'Amount of Dialogue Spoken By Women',x_range = [1930, 2018], y_range = [0, 1.01],
                 tooltips=TOOLTIPS, width=800, height=550, active_scroll='wheel_zoom')

# Create a red to blue color palette
color_mapper = linear_cmap(field_name='proportion_of_dialogue', palette=RdBu[4], low=1.1, high=0)

# Supply inidivudal points values
bokeh_plot.circle(y='proportion_of_dialogue', x=jitter('release_year', width=.2),
         size = 10,
        line_color='black',
        line_alpha=.4,
         source=source,
         color=color_mapper, alpha=.5)

bokeh_plot.title.text_font_size='20pt'

#Make Y axis percentages
bokeh_plot.yaxis.formatter = NumeralTickFormatter(format='0 %')

show(bokeh_plot)

NameError: name 'women_film_df' is not defined